neps/nep-0418.md

```diff
@@ -0,0 +1,90 @@
+Category: Contract
+Created: 18-Oct-2022
+---
+
+## Summary
+
+This proposal is to switch the behavior of the `attached_deposit` host function on the runtime from panicking in view contexts to returning 0. This results in a better devX because instead of having to configure an assertion that there was no attached deposit to a function call only for transactions and not view calls, which is impossible because you can send a transaction to any method, you could just apply this assertion without the runtime aborting in view contexts.
+
+## Motivation
+
+This will allow contract SDK frameworks to add the `attached_deposit == 0` assertion for every function on a contract by default. This behavior matches the Solidity/Eth payable modifier and will ensure that funds aren't sent accidentally to a contract in more cases than currently possible.
+
+This can't be done at a contract level because there is no way of checking if a function call is within view context to call `attached_deposit` conditionally. This means that there is no way of restricting the sending of funds to functions intended to be view only because the abort from within `attached_deposit` can't be caught and ignored from inside the contract.
+
+Initial discussion: https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/attached_deposit.20view.20error
+
+## Rationale and alternatives
+
+The rationale for assigning `0u128` to the pointer (`u64`) passed into `attached_deposit` is that it's the least breaking change. 
+
+The alternative of returning some special value, say `u128::MAX`, is that it would cause some unintended side effects for view calls using the `attached_deposit`. For example, if `attached_deposit` is called within a function, older versions of a contract that do not check the special value will return a result assuming that the attached deposit is `u128::MAX`. This is not a large concern since it would just be a view call, but it might be a bad UX in some edge cases, where returning 0 wouldn't be an issue.
+
+## Specification
+
+After the protocol version, the error inside `attached_deposit` will be skipped, and for all view calls, `0u128` will be set at the pointer passed in.
+
+## Reference Implementation (Required for Protocol Working Group proposals, optional for other categories)
+
+
+Currently, the implementation for `attached_deposit` is as follows:
+```rust
+pub fn attached_deposit(&mut self, balance_ptr: u64) -> Result<()> {
+	self.gas_counter.pay_base(base)?;
+
+	if self.context.is_view() {
+		return Err(HostError::ProhibitedInView {
+			method_name: "attached_deposit".to_string(),
+		}
+		.into());
+	}
+	self.memory_set_u128(balance_ptr, self.context.attached_deposit)
+}
+```
+
+Which would just have to add the check:
+
+```rust
+pub fn attached_deposit(&mut self, balance_ptr: u64) -> Result<()> {
+	self.gas_counter.pay_base(base)?;
+
+	if self.context.protocol_version < VIEW_ATTACHED_DEPOSIT_PROTOCOL_VERSION && self.context.is_view() {
```

Comment:
As @mfornet mentioned, it seems that this proposal does not necessarily trigger a protocol change, which lowers the barrier for this NEP to be approved, I guess.

Reply:
Is that not a side effect we want to avoid? ie someone querying a function at a historical block and it giving inconsistent results back? Do we not have any guarantees about this?

Reply:
This could be considered a client-breaking change update, but it isn't a protocol upgrade. If there were more than one client out there, it could be possible that some clients would implement these features in different ways.

I understand that making this change could break some assumptions about the code that has been written already that relies on current behavior. But even if you do a Protocol Upgrade, it won't solve the issue for them (users can't pin an older protocol version). 

IMO we should use the proposed behavior for legacy blocks as well, but I understand that if this means breaking too many apps, it is unacceptable. In that case, we can apply this behavior after some blocks. Even if the choice is to use this strategy after some fixed block, I still think no Protocol Version update is required (just the client version).

Reply:
hmm, I guess my intuition was that `view` semantics were built into the protocol. I guess that's technically not the case.

I'll change the proposal to not change behaviour based on network upgrade, and we can evaluate and try to minimize the risk of this known unknown and think of other potential risks.